README.md
#       Midnight Snack Build Script

##      What does this do?
This basically automates the process of building Android and its derivatives, as well as uploading the builds to a remote SSH server and sending emails notifying the status of the builds.

1.  Import includes (functions, settings, error descs) and set vars
2.  CheckVariablesExist
3.  LogHeaders
4.  Repo Sync
5.  DeviceLoop
     1. LogHeaders
     2. Lunch
     3. Make
     4. Find and rename zip
     5. MD5sum
     6. Upload zip, then rename
     7. Delete Build zip
6.  TarGZ logs and move to LogFileLoc/archives
7.  Drop email with successes and fails with attached logs (coming in Version 2)

##      Why do I want this?
Because:
  - You're too lazy to do it manually
  - You're too lazy to set up Jenkins

##      How do I use it?

1.  git clone git@github.com:harryyoud/midnightsnack
2.  configure settings in `settings.sh`
3.  ./main.sh
4.  Wait....
5.  Profit (or it failed - check the logs)

##      Configuration
All configuration happens in the settings.sh file. They variables in there are either self-explanatory or have a comment attached

##      Errors
Being a long and complicated drawn out process, the script could fail at lots of points. The various errors you might encounter are explained quite well in errors.sh

##      Improving/Building on top of
I tried to make this as easy as possible to build upon or change, because everybody's needs are different.
Hopefully, the only changes you need to make are in main.sh, to add post-/pre-build hooks or change the order of some things


CHANGELOG.md
v1.2.0:
  - Create remote SSH directory if it doesn't exist when uploading

v1.1.1:
  - Kill jack-server when exiting for errors, otherwise it is left running

v1.1.0:
  - Add option to repopick by topic

v1.0.1:
  - Fixed bad sourcing when main.sh is called from a different directory

v1.0.0:
  - Initial Release
  - All features are ready apart
    - Post-completion mail function on track for next major version
    - Handling make errors more efficiently, by continuing the loop for another device, rather than quit the script


main.sh
#!/bin/bash

# 1.  Import includes (functions, settings, error descs) and set vars
# 2.  CheckVariablesExist
# 3.  LogHeaders
# 4.  Repo Sync
# 5.  DeviceLoop
#       5a. LogHeaders
#       5b. Lunch
#       5c. Make
#       5d. Find and rename zip
#       5e. MD5sum
#       5f. Upload zip, then rename
#       5g. Delete Build zip
# 6.  TarGZ logs
# 7.  Drop email with successes and fails with attached logs (coming in Version 2)



# 1. Import includes (functions, settings, error descs) and set vars
      whereami=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
      source $whereami/functions.sh
      source $whereami/settings.sh
      source $whereami/errors.sh
      GetBuildDate
      StartTimestamp=$(date +%s)
      Version='1.1.0'


# 2.  CheckVariablesExist
#     This is important, as empty variables will wreak havoc
#     We'll only check the "Needed settings"

      CheckVariablesExist
      export USER=$(whoami)
      if [[ UseCcache = true ]]; then
        export USE_CCACHE=1
      else
        export USE_CCACHE=0
      fi

# 3.  LogHeaders
#     We'll output all the admin information at the top of the log file, so it can be seen
#     We'll set the builddate here too, so it's early and can be outputted

      LogMain "Beginning midnightsnack v$Version on $(hostname) for $USER"
      LogMain "The time is $(date +'%T') on $(date +%Y/%m/%d)"
      LogMain "Devices to build for: ${DeviceList[*]}"
      LogMain "Build date is going to be set to $BuildDate"

      if [[ $StopOnWarn = true ]]; then
        LogMain "We'll bail out on warnings"
      else
        LogMain "We'll ignore warnings and bail out on errors"
      fi

      LogMain "Moving into $SourceTreeLoc"
      cd $SourceTreeLoc
      LogMain "Sourcing envsetup.sh"
      LogCommandMainErrors "source build/envsetup.sh"

# 4.  Repo Sync
#     If $SyncOnStart is set, sync repositories
#     Only output errors to log, and send stdout to /dev/null

      if [[ $SyncOnStart = true ]]; then
        LogMain "Syncing repositories"
        LogCommandMainErrors "/home/$USER/bin/repo sync"
      else
        LogMain "\$SyncOnStart not set; skipping repo sync"
      fi
      if ! [[ -z $RepoPicks ]]; then
        LogMain "Applying repopicks from Gerrit (by changeID):"
          for RepoChangeID in "${RepoPicks[@]}"; do
            LogCommandMainErrors "repopick $RepoChangeID"
          done
        LogMain "Repopicks (by changeID) complete"
      else
        LogMain "No repopick changeIDs defined; skipping"
      fi
      if ! [[ -z $RepoTopics ]]; then
        LogMain "Applying repopicks from Gerrit (by topic):"
          for RepoTopic in "${RepoTopics[@]}"; do
            LogCommandMainErrors "repopick -t $RepoTopic"
          done
        LogMain "Repopicks (by topic) complete"
      else
        LogMain "No repopick topics defined; skipping"
      fi


# 5.  DeviceLoop
#     "It's just like replaying the worst day of your life over and over"
#     Begin the mega loop. To summarise, we'll iterate over the DeviceList, say some stuff, build and upload

      LogMain "Beginning main device loop at $(date +'%T')"
      for Device in "${DeviceList[@]}"; do

#       5a. LogHeaders
#       So we'll talk crap to the main log for a minute
        LogMain "Beginning loop as $Device:"

#       5b. Lunch
#       This builds some makefiles and tells mka what device and variant we're building
        LogMain "\tTaking lunch for $Device"
        MidnightSnackLunch $Device

#       5c. Make
#       This kicks off ninja amongst other things that makes the final files for our device
        LogMain "\tKilling Jack"
        LogCommandMainErrors "KillJack"
        LogMain "\tBringing him back to life with more resources"
        LogCommandMainErrors "ResuscitateJack"

        LogMain "\tMaking for $Device. Get the kettle on!"
        MidnightSnackMake
        LogMain "\tMake finished"
        # Take next to last line of makelog, removes ### from start and end and removes colour control characters
        LastLineMakeLogFile=$(tail -n 2 $MakeLogFile | head -n 1 | tr -d \# | sed -r "s/\x1B\[([0-9]{1,2}(;[0-9]{1,2})?)?[mGK]//g" | cut -b 2-)
        LogMain "\t$LastLineMakeLogFile"

#       5d. Find and rename zip
#       This finds the most recently modified zip file in the device out directory, and renames it to something sensible
        GetOutputZip $Device
        LogMain "\tFound zip at $OutputZip"
        GetNewName $Device
        LogMain "\tRenaming it to $NewName"
        NewOutputZip=$SourceTreeLoc/out/target/product/$Device/$NewName
        LogCommandMainErrors "mv $OutputZip $NewOutputZip"

#       5e. MD5SUM
#       Output md5sum of zip to log and file
        GetLocalMD5SUM $NewOutputZip
        LogMain "\tCreating $NewOutputZip.md5sum"
        echo $MD5SUM > $NewOutputZip.md5sum
        LogMain "\tMD5sum of zip: ${MD5SUM:0:32}"

#       5f. Upload zip, then rename
#       We first upload the zip. The zip is named weird, as we don't want people downloading it while it's uploading
#       After both have been uploaded, we can rename the zip
        if [[ $SSHUpload = true ]]; then
          LogMain "\tUploading zip to $SSHHost"
          LogCommandMainErrors "UploadZipAndRename $NewOutputZip $NewName"
          LogMain "\tUploading MD5"
          LogCommandMainErrors "UploadMD5 $NewOutputZip $NewName"
        else
          LogMain "\tSkipping SSH Upload as \$SSHUpload not set"
        fi
#     End it all; I want to die
#     Go back to 5a. and start again
      LogMain "\tFinished main device loop for $Device"
      done
      LogMain "Completed the loop of death. Continuing..."
      LogMain "Killing Jack once and for all"
      LogCommandMainErrors "KillJack"


# 6.  TarGZ Logs
      LogMain "Archiving logs after last message of this script"
      LogMain "Finished script at $(date +'%T') on $(date +%Y/%m/%d)"
      SecondsForScript=$(expr $(date +%s) - $StartTimestamp)
      LogMain "That means it took $(printf '%dh:%dm:%ds\n' $(($SecondsForScript/3600)) $(($SecondsForScript%3600/60)) $(($SecondsForScript%60)))"
      LogMain "And we're done!"
      pushd $LogFileLoc
        LogCommandMainErrors "tar --remove-files -czvf $RomVariant-$RomVersion-$BuildDate-UNOFFICIAL.tar.gz $RomVariant-$RomVersion-$BuildDate-*"
        if ! [[ -e archives ]]; then
          LogCommandMainErrors "mkdir archives"
        fi
        mv $RomVariant-$RomVersion-$BuildDate-UNOFFICIAL.tar.gz archives/
      popd


# 7.  Drop email with successes and fails with attached logs
#     WIP


errors.sh
GetErrorDesc() {
  ErrorNum=$1
  case $ErrorNum in
    200)
      ErrorDesc="Make failed. Check $MakeLogFile"
      ;;
    201)
      ErrorDesc="MidnightSnackLunch() needs 1 argument (the device codename)"
      ;;
    202)
      ErrorDesc="Lunch failed. Is $LunchCommand on the menu?"
      ;;
    211)
      ErrorDesc="GetNewName() needs 1 argument (the device codename)"
      ;;
    212)
      ErrorDesc="Output ZIP was not found (maybe make failed silently?)"
      ;;
    213)
      ErrorDesc="Out folder for $Device doesn't exist. Something weird has happened"
      ;;
    214)
      ErrorDesc="GetOutputZip() needs 1 argument (the device codename)"
      ;;
    215)
      ErrorDesc="Output ZIP provided to GetLocalMD5SUM doesn't exist"
      ;;
    216)
      ErrorDesc="GetLocalMD5SUM() needs 1 argument (the output zip)"
      ;;
    217)
      ErrorDesc="UploadZipAndRename() needs 2 arguments (fullzippath, zipname). You provided only 1"
      ;;
    218)
      ErrorDesc="UploadZipAndRename() needs 2 arguments (fullzippath, zipname). You provided none"
      ;;
    219)
      ErrorDesc="UploadMD5() needs 2 arguments (fullzippath, zipname). You provided only 1"
      ;;
    220)
      ErrorDesc="UploadMD5() needs 2 arguments (fullzippath, zipname). You provided none"
      ;;
    230)
      ErrorDesc="\$SourceTreeLoc not set in settings.sh. This is a required variable"
      ;;
    231)
      ErrorDesc="\$DeviceList is not set in settings.sh. This is a required variable"
      ;;
    232)
      ErrorDesc="\$DeviceList is not an array. This is needs to be an array, regardless of if you have one device to build for"
      ;;
    234)
      ErrorDesc="\$RomVariant is not set in settings.sh. This is a required variable"
      ;;
    235)
      ErrorDesc="\$RomBuildType is not set in settings.sh. This is a required variable"
      ;;
    236)
      ErrorDesc="\$RomVersion is not set in settings.sh. This is a required variable"
      ;;
    238)
      ErrorDesc="\$SSHHost is not set in settings.sh. This is a required variable as \$SSHUpload = true"
      ;;
    239)
      ErrorDesc="\$SSHUser is not set in settings.sh. This is a required variable as \$SSHUpload = true"
      ;;
    240)
      ErrorDesc="\$SSHPort is not set in settings.sh. This is a required variable as \$SSHUpload = true"
      ;;
    241)
      ErrorDesc="\$SSHDirectory is not set in settings.sh. This is a required variable as \$SSHUpload = true"
      ;;
    242)
      ErrorDesc="\$RepoPicks needs to be an array in settings.sh if it is declared"
      ;;
    243)
      ErrorDesc="\$JackRAM is not set in settings.sh. This is a required variable"
      ;;
    244)
      ErrorDesc="\$RepoPicks needs to be an array in settings.sh if it is declared"
      ;;    
    *)
      ErrorDesc="This is embarassing. I have no explanation"
      ;;
  esac;
}


settings.sh
# Needed Settings

    # Log file location - needs to be defined first
    LogFileLoc="/home/harry/android/system/logs"

    SourceTreeLoc="/home/harry/android/system"
    DeviceList=(  "angler"
                  "d2vzw"
                  "deb"
                  "falcon"
                  "flo"
                  "flounder"
                  "flounder_lte"
                # "grouper"
                  "huashan"
                  "klte"
                  "serranoltexx"
                # "tilapia"
                                    )

    # In Settings, Version: 14.1-20161325-UNOFFICIAL-angler
    #   RomVersion is '14.1'
    #   Device is 'angler'
    # In Settings, Build Number: cm_angler-userdebug
    #   RomVariant is 'cm'
    #   RomBuildType is 'userdebug'
    #   Device is 'angler'
    RomVariant="lineage"
    RomBuildType="userdebug"
    RomVersion="14.1"

    # Jack eats RAM
    JackRAM="6g"


# Optionals
    # Set Zip file name to tomorrow's date (date at start of script + 24 hours)
    BuildTomorrow=false
    # Stop script on non-breaking changes
    StopOnWarn=false
    # Sync repo before building
    SyncOnStart=true
    # Use ccache: the superfast cache of previously built files
    UseCcache=true
    DeleteBuildAfterUpload=true
    # If you'd like to pick some specific unmerged changes from Gerrit after a repo sync, here's your chance:
    # RepoPicks is an array so use it in the form below: (Like DeviceList)
    #RepoPicks=( "" )
    # If you'd like to pick an entire topic to pull from Gerrit after a repo sync: (same format as RepoPicks and DeviceList)
    RepoTopics=( "rebrand" )



# SSH Upload
    SSHUpload=true #Therefore define the parameters
      SSHHost=harryyoud.co.uk
      SSHUser=harryyoud
      SSHPort=22
      SSHDirectory=/home/harryyoud/public_html/lineageos/downloads


functions.sh
CheckVariablesExist() {
  if [[ -z $SourceTreeLoc ]]; then
    HandleError 230; fi
  if [[ -z $DeviceList ]]; then
    HandleError 231; fi
  if ! [[ "$(declare -p DeviceList)" =~ "declare -a" ]]; then
    HandleError 232; fi
  if [[ -z $LogFileLoc ]]; then
    HandleError 233; fi
  if [[ -z $RomVariant ]]; then
    HandleError 234; fi
  if [[ -z $RomBuildType ]]; then
    HandleError 235; fi
  if [[ -z $RomVersion ]]; then
    HandleError 236; fi
  if [[ -z $JackRAM ]]; then
    HandleError 243; fi
  if ! [[ -z $RepoPicks ]]; then
    if ! [[ "$(declare -p RepoPicks)" =~ "declare -a" ]]; then
      HandleError 242; fi
  fi
  if ! [[ -z $RepoTopics ]]; then
    if ! [[ "$(declare -p RepoTopics)" =~ "declare -a" ]]; then
      HandleError 244; fi
  fi
  if [[ $SSHUpload = true ]]; then
    if [[ -z $SSHHost ]]; then
      HandleError 238; fi
    if [[ -z $SSHUser ]]; then
      HandleError 239; fi
    if [[ -z $SSHPort ]]; then
      HandleError 240; fi
    if [[ -z $SSHDirectory ]]; then
      HandleError 241; fi
  fi
}

HandleError() {
  # Check we were passed an error code as 1st argument
  if ! [[ -z "$1" ]]; then
    # If it's error 210 (log file not writable), exit without writing to log file (otherwise we go round in circles)
    if [[ $1 = 210 ]]; then
      # Log file not writable
      echo "Log file directory not writable, aborting"
      exit 210
    elif [[ $1 = 233 ]]; then
      # If it's error 233, $LogFileLoc has not been set, so don't write a log
      echo "\$LogFileLoc not defined in settings.sh. This is needed desperately"
      exit 233
    else
      # All other error codes are looked up
      ErrorNum=$1
      # Get error description (listed in ./errors.sh)
      GetErrorDesc $ErrorNum
      LogMain "Error $ErrorNum: $ErrorDesc"
      LogMain 'Killing jack-server and stopping due to error'
      LogCommandMainErrors "KillJack"
      exit $ErrorNum
    fi
  else
    # We weren't passed an error code, so exit
    LogMain "Unspecified Error" "a"
    LogMain 'Killing jack-server and stopping due to error' "a"
    LogCommandMainErrors "KillJack"
    exit 255
  fi
}

HandleWarn() {
  # Check we were passed a warning code as 1st argument
  if ! [[ -z "$1" ]]; then
    WarnNum=$1
    GetErrorDesc $WarnNum
    LogMain "Warning $WarnNum: $ErrorDesc"
    # If StopOnWarn in settings is set, then we stop on non-trivial errors (==warnings)
    if [[ $StopOnWarn = true ]]; then
      LogMain 'Stopping as \$StopOnWarn set' "a"
      exit $WarnNum
    fi
  else
    LogMain "Unspecified Warning (probably non-breaking)" "a"
    if [[ $StopOnWarn = true ]]; then
      LogMain 'Stopping as \$StopOnWarn set' "a"
      exit 255
    fi
  fi
}

LogCommandMake() {
  MakeLogFile=$LogFileLoc'/'$RomVariant'-'$RomVersion'-'$BuildDate'-'UNOFFICIAL'-'$Device'.zip.log'
  # If log file folder isn't writable, error code 210 passed
  if ! [[ -w $LogFileLoc ]]; then
    HandleError 210
  fi
  if ! [[ -z "$2" ]]; then
    # If Log is called in rewrite mode (Log "Blah blah" "r"), overwrite log file
    if [[ "$2" = "r" ]]; then
      $1 >  "$MakeLogFile" 2>&1
                #          ^^^^ redirect errors too
    # Otherwise, append
    else
      $1 >>  "$MakeLogFile" 2>&1
    fi
  else
    $1 >>  "$MakeLogFile" 2>&1
  fi;
}

LogMake() {
  MakeLogFile=$LogFileLoc'/'$RomVariant'-'$RomVersion'-'$BuildDate'-'UNOFFICIAL'-'$Device'.zip.log'
  # If log file folder isn't writable, error code 210 passed
  if ! [[ -w $LogFileLoc ]]; then
    HandleError 210
  fi
  if ! [[ -z "$2" ]]; then
    # If Log is called in rewrite mode (Log "Blah blah" "r"), overwrite log file
    if [[ "$2" = "r" ]]; then
      printf "%b\r\n"  "$1" >  "$MakeLogFile" 2>&1
                     #          ^^^^ redirect errors too
    # Otherwise, append
    else
      printf "%b\r\n"  "$1" >>  "$MakeLogFile" 2>&1
    fi
  else
    printf "%b\r\n"  "$1" >>  "$MakeLogFile" 2>&1
  fi;
}

LogCommandMain() {
  MainLogFile=$LogFileLoc'/'$RomVariant'-'$RomVersion'-'$BuildDate'-'UNOFFICIAL'.log'
  # If log file folder isn't writable, error code 210 passed
  if ! [[ -w $LogFileLoc ]]; then
    HandleError 210
  fi
  if ! [[ -z "$2" ]]; then
    # If Log is called in rewrite mode (Log "Blah blah" "r"), overwrite log file
    if [[ "$2" = "r" ]]; then
      $1 >  "$MainLogFile" 2>&1
                #          ^^^^ redirect errors too
    # Otherwise, append
    else
      $1 >>  "$MainLogFile" 2>&1
    fi
  else
    $1 >>  "$MainLogFile" 2>&1
  fi;
}

LogCommandMainErrors() {
  MainLogFile=$LogFileLoc'/'$RomVariant'-'$RomVersion'-'$BuildDate'-'UNOFFICIAL'.log'
  # If log file folder isn't writable, error code 210 passed
  if ! [[ -w $LogFileLoc ]]; then
    HandleError 210
  fi
  if ! [[ -z "$2" ]]; then
    # If Log is called in rewrite mode (Log "Blah blah" "r"), overwrite log file
    if [[ "$2" = "r" ]]; then
      $1 >/dev/null 2> "$MainLogFile"
      #             ^^^^^^^^^^^^^^^^^ redirect errors to MainLog
      #  ^^^^^^^^^^ discard stdout
    # Otherwise, append
    else
      $1 >>/dev/null 2>> "$MainLogFile"
    fi
  else
    $1 >>/dev/null 2>> "$MainLogFile"
  fi;
}

LogMain() {
  MainLogFile=$LogFileLoc'/'$RomVariant'-'$RomVersion'-'$BuildDate'-'UNOFFICIAL'.log'
  # If log file folder isn't writable, error code 210 passed
  if ! [[ -w $LogFileLoc ]]; then
    HandleError 210
  fi
  if ! [[ -z "$2" ]]; then
  # If log file folder isn't writable, error code 210 passed
    if [[ "$2" = "r" ]]; then
      printf "%b\r\n" "$1" > "$MainLogFile" 2>&1
    # Otherwise, append
    else
      printf "%b\r\n" "$1" >> "$MainLogFile" 2>&1
    fi
  else
    printf "%b\r\n" "$1" >> "$MainLogFile" 2>&1
  fi;
}

MidnightSnackLunch() {
  # Check we've been given argument 1 (device)
  if ! [[ -z $1 ]]; then
    # Run this from source directory...
    cd $SourceTreeLoc
    # ... for envsetup.sh to work
    LogCommandMainErrors "source build/envsetup.sh"
    LunchCommand=$RomVariant'_'$1'-'$RomBuildType
    LogCommandMake "lunch $LunchCommand" || HandleError 202
  else
    # Gimme more arguments
    HandleError 201
  fi
}

MidnightSnackMake() {
  LogCommandMake "mka otapackage" || HandleError 200
}

GetBuildDate() {
  if [[ $BuildTomorrow = true ]]; then
      # Get YYYYMMDD for tomorrow
      BuildDate=$(date --date="+1 day" +%Y%m%d);
  else
    # Get YYYYMMDD for today
    BuildDate=$(date +%Y%m%d);
  fi
}

GetNewName() {
  # Check we've been given the first argument (device)
  if ! [[ -z $1 ]]; then
    NewName=$RomVariant'-'$RomVersion'-'$BuildDate'-'UNOFFICIAL'-'$1'.zip'
  else
    # Can I haz moar argument?
    HandleError 211
  fi
}

GetOutputZip() {
  # Check we've been given first argument (device)
  if ! [[ -z $1 ]]; then
    # Check device output folder exists
    if [[ -e $SourceTreeLoc/out/target/product/$1 ]]; then
                # find *.zip in the root of the output directory, reverse ordered by date modified, take the top line
      OutputZip=$(find $SourceTreeLoc/out/target/product/$1/ -maxdepth 1 -name '*.zip' -printf "%T+\t%p\n" | sort -r | cut -f 2- | head -n 1)
                #                                                                                                                  ^^^^^^^^^ take top line
                #                                                                                                      ^^^^^^^^^ take the second field after the tab (chop date modified off front)
                #                                                                                            ^^^^^^^ sort numerically in reverse
                #                                                                      ^^^^^^^^^^^^^^^^^^^ print as "2017-01-02+18:45:41.7878729150 android/system/out/target/product/angler/cm_angler-ota-a0db5d5712.zip"
                #                                                        ^^^^^^^^^^^^^ All zip files
                #                                            ^^^^^^^^^^^ Only files in the root of the directory
      # If find found a zip, output it
      if [[ -z $OutputZip ]]; then
        HandleError 212
      fi
    else
      # The device output folder doesn't exist
      HandleError 213
    fi
  else
    # Not given device argument
    HandleError 214
  fi
}

GetLocalMD5SUM() {
  # Check we've been given first argument (OutputZip)
  if ! [[ -z $1 ]]; then
    # Check output zip exists
    if [[ -e $1 ]]; then
      MD5SUM=$(md5sum $1)
    else
      # Output file doesn't exist
      HandleError 215
    fi
  else
    # First argument not given
    HandleError 216
  fi
}

UploadZipAndRename() {
  # Check we've been given first argument (Absolute path to zip)
  if ! [[ -z $1 ]]; then
    # Check we've been given second argument (Name of zip file)
    if ! [[ -z $2 ]]; then
      LocalZipPath=$1
      LocalZipName=$2
      ssh $SSHUser@$SSHHost -p $SSHPort "mkdir -p $SSHDirectory/$Device"
      # Upload Zip file to nameofzipfile.zip.part
      scp -P $SSHPort $LocalZipPath $SSHUser@$SSHHost:"$SSHDirectory/$Device/$LocalZipName.part"
      # Move nameofzipfile.zip.part to nameofzipfile
      ssh $SSHUser@$SSHHost -p $SSHPort "mv $SSHDirectory/$Device/$LocalZipName.part $SSHDirectory/$Device/$LocalZipName"
    else
      # Second argument not given
      HandleError 217
    fi
  else
    # First argument not given
    HandleError 218
fi
}

UploadMD5() {
  # Check for first argument (Absolute path to zip file)
  if ! [[ -z $1 ]]; then
    # Check for second argument (Name of zip file)
    if ! [[ -z $2 ]]; then
      LocalZipPath=$1
      LocalZipName=$2
      ssh $SSHUser@$SSHHost -p $SSHPort "mkdir -p $SSHDirectory/$Device" 
      scp -P $SSHPort $LocalZipPath.md5sum $SSHUser@$SSHHost:"$SSHDirectory/$Device/$LocalZipName.md5sum"
    else
      # Second argument not given
      HandleError 219
    fi
  else
    # First argument given
    HandleError 220
fi
}

KillJack() {
  # Kill the jack-server so we can restart with more RAM
  cd $SourceTreeLoc
  ./prebuilts/sdk/tools/jack-admin kill-server
}

ResuscitateJack() {
  # Bring Jack back with more RAM
  cd $SourceTreeLoc
  export JACK_SERVER_VM_ARGUMENTS="-Dfile.encoding=UTF-8 -XX:+TieredCompilation -Xmx$JackRAM"
  ./prebuilts/sdk/tools/jack-admin start-server
}
